flink-tests/src/test/java/org/apache/flink/test/checkpointing/RegionFailoverITCase.java

```diff
@@ -0,0 +1,307 @@
+		env.disableOperatorChaining();
+		env.getConfig().setRestartStrategy(RestartStrategies.fixedDelayRestart(NUM_OF_RESTARTS, 0L));
+		env.getConfig().disableSysoutLogging();
+
+		// there exists num of 'NUM_OF_REGIONS' individual regions.
+		env.addSource(new StringGeneratingSourceFunction(NUM_ELEMENTS, NUM_ELEMENTS / NUM_OF_RESTARTS))
+			.setParallelism(NUM_OF_REGIONS)
+			.map(new FailingMapperFunction(NUM_OF_RESTARTS))
+			.setParallelism(NUM_OF_REGIONS);
+
+		// another stream graph totally disconnected with the above one.
+		env.addSource(new StringGeneratingSourceFunction(NUM_ELEMENTS, NUM_ELEMENTS / NUM_OF_RESTARTS)).setParallelism(1)
+			.map((MapFunction<Integer, Object>) value -> value).setParallelism(1);
+
+		return env.getStreamGraph().getJobGraph();
+	}
+
+	private static class StringGeneratingSourceFunction extends RichParallelSourceFunction<Integer>
+		implements CheckpointListener, CheckpointedFunction {
+		private static final long serialVersionUID = 1L;
+
+		private final long numElements;
+		private final long checkpointLatestAt;
+
+		private int index = -1;
+
+		private int lastRegionIndex = -1;
+
+		private volatile boolean isRunning = true;
+
+		private ListState<Integer> listState;
+
+		private static final ListStateDescriptor<Integer> stateDescriptor = new ListStateDescriptor<>("list-1", Integer.class);
+
+		private ListState<Integer> unionListState;
+
+		private static final ListStateDescriptor<Integer> unionStateDescriptor = new ListStateDescriptor<>("list-2", Integer.class);
+
+		StringGeneratingSourceFunction(long numElements, long checkpointLatestAt) {
+			this.numElements = numElements;
+			this.checkpointLatestAt = checkpointLatestAt;
+		}
+
+		@Override
+		public void run(SourceContext<Integer> ctx) throws Exception {
+			if (index < 0) {
+				// not been restored, so initialize
+				index = 0;
+			}
+
+			while (isRunning && index < numElements) {
+				//noinspection SynchronizationOnLocalVariableOrMethodParameter
+				synchronized (ctx.getCheckpointLock()) {
+					index += 1;
+					ctx.collect(index);
+				}
+
+				if (numCompletedCheckpoints < 3) {
+					// not yet completed enough checkpoints, so slow down
+					if (index < checkpointLatestAt) {
+						// mild slow down
+						Thread.sleep(1);
+					} else {
+						// wait until the checkpoints are completed
+						while (isRunning && numCompletedCheckpoints < 3) {
+							Thread.sleep(300);
+						}
+					}
+				}
+				if (jobFailedCnt.get() < NUM_OF_RESTARTS) {
+					// slow down if job has not failed for 'NUM_OF_RESTARTS' times.
+					Thread.sleep(1);
+				}
+			}
+		}
+
+		@Override
+		public void cancel() {
+			isRunning = false;
+		}
+
+		@Override
+		public void notifyCheckpointComplete(long checkpointId) throws Exception {
+			if (getRuntimeContext().getIndexOfThisSubtask() == NUM_OF_REGIONS - 1) {
+				lastCompletedCheckpointId = checkpointId;
+				snapshotIndicesOfSubTask.put(checkpointId, lastRegionIndex);
+				numCompletedCheckpoints++;
```

Comment:
You are using a non-atomic operation on a `volatile` field. Maybe the field is not required to be volatile or if it is truly required you should rather use an `AtomicInteger` to avoid rare failures.